runtime/app.go

```diff
@@ -0,0 +1,152 @@
+	baseAppOptions      []BaseAppOption
+	txHandler           tx.Handler
+	msgServiceRegistrar grpc.Server
+}
+
+// RegisterModules registers the provided modules with the module manager and
+// the basic module manager. This is the primary hook for integrating with
+// modules which are not registered using the app config.
+func (a *App) RegisterModules(modules ...module.AppModule) error {
+	for _, appModule := range modules {
+		name := appModule.Name()
+		if _, ok := a.ModuleManager.Modules[name]; ok {
+			return fmt.Errorf("AppModule named %q already exists", name)
+		}
+		a.ModuleManager.Modules[name] = appModule
+
+		if _, ok := a.basicManager[name]; ok {
+			return fmt.Errorf("AppModuleBasic named %q already exists", name)
+		}
+
+		a.basicManager[name] = appModule
+		appModule.RegisterInterfaces(a.interfaceRegistry)
+		appModule.RegisterLegacyAminoCodec(a.amino)
+
+	}
+	return nil
+}
+
+// Load finishes all initialization operations and loads the app.
+func (a *App) Load(loadLatest bool) error {
+	if a.msgServiceRegistrar != nil {
+		configurator := module.NewConfigurator(a.cdc, a.msgServiceRegistrar, a.GRPCQueryRouter())
+		a.ModuleManager.RegisterServices(configurator)
+	}
+
+	if len(a.config.InitGenesis) != 0 {
+		a.ModuleManager.SetOrderInitGenesis(a.config.InitGenesis...)
+		a.SetInitChainer(a.InitChainer)
+	}
+
+	if len(a.config.BeginBlockers) != 0 {
+		a.ModuleManager.SetOrderBeginBlockers(a.config.BeginBlockers...)
+		a.SetBeginBlocker(a.BeginBlocker)
+	}
+
+	if len(a.config.EndBlockers) != 0 {
+		a.ModuleManager.SetOrderEndBlockers(a.config.EndBlockers...)
+		a.SetEndBlocker(a.EndBlocker)
+	}
+
+	if loadLatest {
+		if err := a.LoadLatestVersion(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// BeginBlocker application updates every begin block
+func (app *App) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) abci.ResponseBeginBlock {
+	return app.ModuleManager.BeginBlock(ctx, req)
+}
+
+// EndBlocker application updates every end block
+func (app *App) EndBlocker(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
+	return app.ModuleManager.EndBlock(ctx, req)
+}
+
+// InitChainer initializes the chain.
+func (a *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.ResponseInitChain {
+	var genesisState map[string]json.RawMessage
+	if err := json.Unmarshal(req.AppStateBytes, &genesisState); err != nil {
+		panic(err)
+	}
+	return a.ModuleManager.InitGenesis(ctx, a.cdc, genesisState)
+}
+
+// RegisterAPIRoutes registers all application module routes with the provided
+// API server.
+func (a *App) RegisterAPIRoutes(apiSvr *api.Server, _ config.APIConfig) {
+	clientCtx := apiSvr.ClientCtx
+	// Register new tx routes from grpc-gateway.
+	authtx.RegisterGRPCGatewayRoutes(clientCtx, apiSvr.GRPCGatewayRouter)
+	// Register new tendermint queries routes from grpc-gateway.
+	tmservice.RegisterGRPCGatewayRoutes(clientCtx, apiSvr.GRPCGatewayRouter)
+
```

Comment:
Recognizing these 2 `RegisterGRPCGatewayRoutes` calls were ported from `SimApp` initialization to here, but why the need for special casing their initialization outside of the `BasicManager` below? Any further reading on these special cased modules?

Reply:
These really should have been in a dedicated module before. At this point runtime seems like a somewhat reasonable place to put them, although I'm not sure why the authtx one isn't just in auth